Download resume for s3 datastore

giggsoff opened the pull request:
This PR allow us to keep information about downloaded parts for s3 datastore in memory and resume download from the previous try.

eriknordmark commented on pkg/pillar/cmd/downloader/syncop.go:
```diff
@@ -44,14 +44,15 @@ func handleSyncOp(ctx *downloaderContext, key string,
 			retryTime, config.Name, err)
 		handleSyncOpResponse(ctx, config, status, locFilename, key,
-			errStr, cancelled)
+			errStr, cancelled, true)
```
Is the last added argument really cleanOnError?

eriknordmark reviewed (commented):
Some nits but my fundamental questions is around the requirements for this feature. Maybe we should have a bit of a requirements discussion?

What you have implemented ensures that if at least one 5Mbyte block has been downloaded, then that block will not be re-downloaded as long as the device is not powered off. But if only 4Mbyte can be downloaded before the network drops out for a bit so the connection times out, those 4Mbyte are discarded and we start from zero.

Also, if 999Mbyte of a 1Gbyte image has been downloaded and then the user powers off the machine for the night, hence the device running EVE-OS is powered off, then when it is powered back on we restart from zero. 

It isn't clear to me that this is sufficient. It certainly would be nice to be able to determine how much was downloaded before the error and use all of those bytes when restarting the download. And even better if on boot the downloader could detect that there is a partial download in /persist/vault/downloader/pending/ and use those bytes as a starting point.

giggsoff commented:
> What you have implemented ensures that if at least one 5Mbyte block has been downloaded, then that block will not be re-downloaded as long as the device is not powered off. But if only 4Mbyte can be downloaded before the network drops out for a bit so the connection times out, those 4Mbyte are discarded and we start from zero.

Correct. I can try to save not only part index but also bytes written to the part, it looks more complicated. Do we really need to keep max 25MB (part_size*concurrence) here?

> Also, if 999Mbyte of a 1Gbyte image has been downloaded and then the user powers off the machine for the night, hence the device running EVE-OS is powered off, then when it is powered back on we restart from zero.

Reworked to keep data across reboots. I modified logic for file name to be some hash on datastoreid+url+size to keep it across reboots (for now it generates randomly). Also added file with `.progress` in the end to keep information about downloaded parts across reboots.

rvs commented on libs/zedUpload/awsutil/s3api.go:
```diff
@@ -18,8 +18,12 @@ import (
+	//S3Concurrency parallels parts download limit
+	S3Concurrency = 5
+	//S3PartSize size of part to download
+	S3PartSize = 5 * 1024 * 1024
```
Tiny nit -- but would it make more sense to say `S3PartSize = S3Concurrency * 1024 * 1024` or is not not implied?

giggsoff replied:
Well, they are separate. We run 5 routines, they all download their part

rvs reviewed (commented):
Can you please separate vendoring into pillar into its own commit @giggsoff ?

Otherwise structurally looks fine to me.

eriknordmark commented on pkg/pillar/types/downloadertypes.go:
```diff
@@ -224,3 +228,20 @@ type DatastoreContext struct {
+
+// DownloadedParts keeps information about downloaded parts of blob
+type DownloadedParts struct {
+	PartSize int64           // not used now, but may be used when partsize will be variable
```
What's the intended meaning? Size of one partition? Total size? 
Does it make sense introducing it if its use isn't well-defined and commented?

giggsoff replied:
I put here size of part (maximum size). It may be useful if we will decide to change part size in the future (and will do some kind of re-calculation what we need to download with new part sizes).

eriknordmark replied:
OK so then the comment should be modified since it is used in the sense that it it set.
Is this a correct description?
// Set to the maximum partition size

giggsoff replied:
Modified, thanks for the comment

eriknordmark commented on libs/zedUpload/go.mod:
```diff
@@ -18,3 +18,5 @@ require (
 
 replace github.com/lf-edge/eve/api/go => ../../api/go
+
+replace github.com/lf-edge/eve/pkg/pillar => ../../pkg/pillar
```
Why do we need this? The dependency is from pillar to zedUpload and here it looks like it is the other way around. Are you using some definitions from pillar?

eriknordmark commented on libs/zedUpload/awsutil/s3sync.go:
```diff
@@ -8,24 +8,28 @@ import (
 	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/service/s3"
 	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/lf-edge/eve/pkg/pillar/types"
```
Please define the struct (is it downloadedparts?) in zedupload and import that into pillar.

eriknordmark reviewed (changes requested):
Define DownloadedParts in zedUpload instead of in pillar and then we'll run this through the tests.

eriknordmark reviewed (approved):
Thanks - LGTM

eriknordmark reviewed (approved):
Run eden again

eriknordmark reviewed (approved):
Run eden again
